src/density_interface.jl

```diff
@@ -0,0 +1,47 @@
+@inline DensityInterface.hasdensity(::Distribution) = true
+
+for (di_func, d_func) in ((:logdensityof, :logpdf), (:densityof, :pdf))
+    @eval begin
+        DensityInterface.$di_func(d::Distribution, x) = $d_func(d, x)
+
+        DensityInterface.$di_func(d::Distribution{ArrayLikeVariate{N}}, x::AbstractArray{<:Real,N}) where N = $d_func(d, x)
+
+        function DensityInterface.$di_func(d::Distribution{ArrayLikeVariate{N}}, x::AbstractArray{<:Real}) where N
+            throw(ArgumentError("$(DensityInterface.$di_func) doesn't support multiple samples as an argument"))
+        end
+
+        function DensityInterface.$di_func(d::Distribution{ArrayLikeVariate{N}}, x::AbstractVector{<:AbstractArray{<:Real}}) where N
+            throw(ArgumentError("$(DensityInterface.$di_func) doesn't support multiple samples as an argument"))
+        end
+    end
+end
+
+
+"""
+    IIDDensity(distribution::Distribution)
+
+Represents the probability density of `d` and implicit product distributions
+over d.
+
+Use `DensityInterface.logdensityof(d, x)` to compute the logarithm density
+value at `x`. `x` may be a single variate of `d` or a whole set of variates
+of `d`.
+
+If `x` is a single variate of `d`, the density is the PDF of `d`.
+
+If `x` is a set of variates (given as a higher-dimensional array or
+and array of arrays), the density is the PDF of an implicit product
+distribution over `d`, the size of the product is implied by the size of
+the set.
+
+`DensityInterface.logdensityof(d, x)` is equivalent to `loglikelihood(d, x)`.
```

Comment:
Can you move this to the definition of `logdensityof(::IIDDensity, x)`? This would it also make clearer what `d` is, I assume.

Reply:
Do we really want to append all of that to the function documentation of `logdensityof`?

Reply:
I have improved the docstring, to make it clearer what `d` is.

I don't think every type that implements DensityInterface should add to the docstring of `logdensityof` though, otherwise it could be come excessively long.

Reply:
Yes, one should only add a docstring for `logdensityof` if it differs from the default behaviour, i.e., is special in some way. I think this is the case for `IIDDensity`, in contrast to a standard `Distribution`, but it's a matter of taste.

One can filter docstrings in the REPL if there are multiple ones for e.g. a function such as `any` by the call syntax and the types of the arguments:
```julia
julia> d = Uniform(); x = 0.3;

julia>? logdensityof(d, x) # should only display docstring(s) that match `logdensityof(::Uniform, ::Float64)`

julia>? logdensityof(d) # should only display docstring(s) that match `logdensityof(::Uniform)`
```

Reply:
Sure, but I usually type `?somefunction` first. :-)

As you noted, use of `IIDDensity` may be limited to a few applications for now, which is why we're not exporting it. So I would prefer not to add IIDDensity to the docstring of `logdensityof`. Ok with you?

Reply:
The behavior of `IIDDensity` is indeed a bit "special", and I don't want to "advertise" it too much as something people should imitate when implementing `logdensityof`. ;-)

Reply:
I assume other packages will also want to add docstrings to `logdensityof` in some cases, to explain specific assumptions or behaviour (e.g., AbstractPPL will also add a docstring). IMO this is reasonable and fine if it additional explanations are required, e.g., to clarify which density is evaluated.

I agree though that it is a bit weird to extend the documentation of `logdensityof` with the `IIDDensity` case as long as `IIDDensity` is not exported. I suggest keeping it in the documentation of the type for now. However, I think this should be changed when/if `IIDDensity` is exported, so maybe it would be good to add a `TODO:` note as a comment.

Reply:
I agree.

> maybe it would be good to add a TODO: note as a comment.

Done.